compare/compare.c

```diff
@@ -442,18 +444,54 @@ static bool add_img_item(ItemList *list, const char *file, int len)
     if (!item->name)
         return out_of_memory();
     item->name[pos] = '\0';
+    item->dir = dir;
     item->time = 0;
     for (int i = first; i < len; i++)
         item->time = 10 * item->time + (file[i] - '0');
     list->n_items++;
     return true;
 }
 
+static bool process_dir(ItemList *list, DIR *dir, const char *path, ASS_Library *lib)
+{
+    while (true) {
+        struct dirent *file = readdir(dir);
+        if (!file)
+            break;
+        const char *name = file->d_name;
+        if (name[0] == '.')
+            continue;
+        const char *ext = strrchr(name + 1, '.');
+        if (!ext)
+            continue;
+
+        if (!strcmp(ext, ".png")) {
+            if (add_img_item(list, path, name, ext - name))
+                continue;
+        } else if (!lib) {
+            continue;
+        } else if (!strcmp(ext, ".ass")) {
+            if (add_sub_item(list, name, ext - name))
+                continue;
+        } else if (!strcmp(ext, ".ttf") || !strcmp(ext, ".otf") || !strcmp(ext, ".pfb")) {
+            if (load_font(lib, path, name))
+                continue;
+            printf("Cannot load font '%s'!\n", name);
+        } else {
+            continue;
+        }
+        closedir(dir);
+        return false;
+    }
+    closedir(dir);
+    return true;
```

Comment:
I find this bit somewhat confusing, the first `closedir; return false;` is dead code after all.
Why not declare `file` first outside of the loop and then just `while(file = readdir(dir))`?

`readdir` is not part of ISO-C99 btw, though for an optional utility like `compare` this probably doesn't matter too much.

I now see `readdir` is also used in `ass_fontselect`...

Reply:
It's not dead code but error handling branch. It's executed in case of `add_*_item()` or `load_font()` failure. I can replace it with the usual `goto fail;` trick if it's too hard to understand.

Reply:
oups, my bad. guess I shouldn't do code reviews late at night.
`goto fail` may be indeed a bit easier to read, but *(now that I'm no longer half-asleep)* the current approach also seems ok.

To elaborate on the `readdir`-part: As it's also used in the lib this isn't a problem. Though even if it wasn't, I wouldn't be opposed to allow using all POSIX-extensions (so also eg `strcasecmp`) in separate, optional utilities like `compare`.
It being used in the core-lib, which is otherwise mostly ISO-C, was a bit surprising to me, but there is no sane way to iterate a directory in pure ISO-C afaik.

Reply:
I think it's mostly safe to use POSIX functions. And for OSes without (complete) POSIX implementation (Windows) we can use `ass_compat.h` with fallbacks.